lib/Sema/CSStep.cpp

```diff
@@ -783,10 +783,46 @@ bool DisjunctionStep::attempt(const DisjunctionChoice &choice) {
 
 bool ConjunctionStep::attempt(const ConjunctionElement &element) {
   ++CS.solverState->NumConjunctionTerms;
+
+  // Outside or previous element score doesn't affect
+  // subsequent elements.
+  CS.solverState->BestScore.reset();
+
+  // Apply solution inferred for all the previous elements
+  // because this element could reference declarations
+  // established in previous element(s).
+  if (!Solutions.empty()) {
+    assert(Solutions.size() == 1);
+    // Note that solution is removed here. This is done
+    // because we want build a single complete solution
+    // incrementally.
+    CS.applySolution(Solutions.pop_back_val());
+  }
+
+  // Bring all of the referenced variables into scope.
+  for (auto *typeVar : element.getReferencedVars())
+    CS.addTypeVariable(typeVar);
+
+  // Make sure that element is solved in isolation
+  // by dropping all scoring information.
```

Comment:
Because score from the outside is unrelated to the elements in the body, solutions can't be ranked on score from context and score from other elements because they are standalone units for type-checker, only types are transferable between them.